Platform/OpenNtfsDxe/Disc.c

```diff
@@ -0,0 +1,747 @@
+      if ((LRecord->Type == Type) || (Type == 0)) {
+        break;
+      }
+
+      if (BufferSize < LRecord->RecordLength) {
+        DEBUG ((DEBUG_INFO, "NTFS: (FindAttr #2) $ATTRIBUTE_LIST is corrupted.\n"));
+        return NULL;
+      }
+
+      if (LRecord->RecordLength == 0) {
+        DEBUG ((DEBUG_INFO, "NTFS: (FindAttr #3) $ATTRIBUTE_LIST is corrupted.\n"));
+        return NULL;
+      }
+
+      Attr->Next += LRecord->RecordLength;
+      BufferSize -= LRecord->RecordLength;
+      LRecord = (ATTR_LIST_RECORD *) Attr->Next;
+    }
+
+    if ((Attr->Next + sizeof (ATTR_LIST_RECORD)) >= Attr->Last) {
+      return NULL;
+    }
+
+    // if ((Attr->Flags & NTFS_AF_MFT_FILE) && (Type == AT_DATA)) {
+    //   Attr->Flags |= NTFS_AF_GPOS;
+    //   Attr->Current = Attr->Next;
+    //   AttrStart = Attr->Current;
+    //
+    //   if (BufferSize >= 0x18) {
+    //     *(UINT32 *) (AttrStart + 0x10) = (UINT32)(Attr->BaseMftRecord->File->FileSystem->FirstMftRecord / mSectorSize);
+    //     *(UINT32 *) (AttrStart + 0x14) = (UINT32)(Attr->BaseMftRecord->File->FileSystem->FirstMftRecord / mSectorSize + 1);
+    //   } else {
+    //     DEBUG ((DEBUG_INFO, "NTFS: (FindAttr #7) File record is corrupted.\n"));
+    //     return NULL;
+    //   }
+    //
+    //   AttrStart = Attr->Next + ReadUnaligned16((UINT16 *)(AttrStart + 0x04));
+    //   while ((AttrStart + sizeof (UINT32) + sizeof (UINT16)) < Attr->Last) {
+    //     if (*(UINT32 *)AttrStart != Type) {
+    //       break;
+    //     }
+    //
+    //     Status = ReadAttr (
+    //       Attr,
+    //       AttrStart + 0x10,
+    //       ReadUnaligned32((UINT32 *)(AttrStart + 0x10)) * mFileRecordSize,
+    //       mFileRecordSize
+    //       );
+    //     if (EFI_ERROR (Status)) {
+    //       return NULL;
+    //     }
+    //
+    //     if (ReadUnaligned16((UINT16 *)(AttrStart + 4)) == 0) {
+    //       DEBUG ((DEBUG_INFO, "NTFS: (FindAttr #8) File record is corrupted.\n"));
+    //       return NULL;
+    //     }
+    //
+    //     AttrStart += ReadUnaligned16((UINT16 *)(AttrStart + 4));
+    //   }
+    //   Attr->Next = Attr->Current;
+    //   Attr->Flags &= ~NTFS_AF_GPOS;
+    // }
+
+    goto retry;
+  }
+
+  return NULL;
+}
+
+EFI_STATUS
+EFIAPI
+InitFile (
+  IN OUT NTFS_FILE  *File,
+  IN     UINT64     RecordNumber
+  )
+{
+  EFI_STATUS         Status;
+  ATTR_HEADER_RES    *Attr;
+  FILE_RECORD_HEADER *Record;
+
+  File->InodeRead = TRUE;
+
+  File->FileRecord = AllocateZeroPool (mFileRecordSize);
+  if (File->FileRecord == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Status = ReadMftRecord (File->File, File->FileRecord, RecordNumber);
+  if (EFI_ERROR (Status)) {
+    FreePool (File->FileRecord);
+    return Status;
+  }
+
+  Record = (FILE_RECORD_HEADER *) File->FileRecord;
+
+  if ((Record->Flags & IS_IN_USE) == 0) {
+    DEBUG ((DEBUG_INFO, "NTFS: MFT Record 0x%Lx is not in use\n", RecordNumber));
```

Comment:
For UINT64 it should be `%LLx`.

Reply:
No such thing in EDK II. Should be left as is.

Reply:
grep over UDK showed:
`%LLx`: 0
`%LLX`: 1
`%LX`:  16
`%Lx`:  139